### ISSUE TITLE
Request:  line drawing capability

### ISSUE THREAD

**Issue by Alexustas** [MOARdV/AvionicsSystems#17]
Line with given Start and End points, with adjustable width  (if possible), positive/negative/ zero color and variable, "oneShot" option, variable for show/ don't show, refreshRate

As the possible options:
**"Multi-line"**
which is drawn based on a given set of points. I think 10 points will be more than enough. You probably already guessed that in the first place, this feature will need to draw a route for the selected waypoint

**"Styles"**
possibly different rendering styles, such as dashed, dotted, double line. I do not know what capabilities can provide for this Unity

**"Predefined primitives"**
such as a rectangle (a special case of multi-line), circle (center, radius)

**Comment by MOARdV** [comment]
Unity LineRenderer supports width (and be different at the start and end) and color (also can be different at start and end).  It does not support styles exactly, but it can have a texture (so maybe textures for different styles?).  I do not know exactly how the textures work - I have never tried it.

**Comment by Alexustas** [comment]
>but it can have a texture (so maybe textures for different styles?)

if you can tell me any specifications for the textures, I draw them

**Comment by MOARdV** [comment]
> if you can tell me any specifications for the textures, I draw them

When I can start experimenting, I will get you specifications.  Too many things to do all at the same time right now. :)

**Comment by Alexustas** [comment]
at the moment I only need to know that it is quite possible to do in the future, so that I can start thinking about the design of the new devices and new features. I do not expect that you will do everything right now ))

**Comment by DeputyLOL** [comment]
Do the lines have to have a texture or can they just be a plain coloured line?
I could use this as a vertical speed gauge on the PFD.
I'm guessing Alexustas is gonna use it as a map/navigation.... thing.

**Comment by MOARdV** [comment]
> Do the lines have to have a texture or can they just be a plain coloured line?

They can always use the default solid texture, which would make them a plain colored line (or even coloured line, on your side of the pond :) ).

**Comment by DeputyLOL** [comment]
Splendid my good man. 
Care for some tea? Its my favoUrite!

**Comment by MOARdV** [comment]
Okay, I have started on this, but it probably will not be ready before this weekend.  Allowing colors and vertices to be variable is complex, and I need to decide how to make it work.  However, once I can make it work, I could allow variable color in other nodes, maybe.

Here is what I have so far for the node:
```
LINE_STRING
{
	name = Line String renderer
		
	position = 256, 256

	startColor = 0, 255, 0, 255
	endColor = 255, 255, 0, 255
		
	startWidth = 1
	endWidth = 4
		
	vertex = 0, 0
	vertex = 64, 0
	...
		
	variable = 1
	range = 0.5, 1.5
}
```

* **name** - same as all other nodes.
* **variable**, **range** - optional enable / disable variable, like other nodes.
* **position** - center of the line drawing region.
* **startColor** - required, color for the "start" of the line.  All four numbers can be variables (ie, `fc.GetThrottle() * 255`).
* **endColor** - optional - uses 'startColor' if it is missing.  Color for the "end" of the line.
* **startWidth** - required, can be variable.  Width at the start of the line.
* **endWidth** - optional - uses 'startWidth' if it is missing.  Width at the end of the line.
* **vertex** - 2 required, more optional.  (x, y) displacement from 'position' for each vertex.  Both coordinates can be variables.

Draws a line string (multiple vertices).  To make a closed polygon, like a rectangle, requires having the first and last vertex the same.

Predefined primitives will be a later addition, although a rectangle is easy with this approach (circle requires more maths, so it will be easier to make a special node for it).

**Comment by MOARdV** [comment]
LINE_STRING is now ready for use.  Dev build on [DropBox](https://www.dropbox.com/s/5lrxoiyjk96su34/AvionicsSystems-0.11.0b1.zip?dl=0).  Official documentation on the [wiki](https://github.com/MOARdV/AvionicsSystems/wiki/MASMonitor#line_string).

One thing I did not add, but I could, is a rotateVariable that would rotate all of the line strings around 'position', so all of the lines could rotate without having to recompute vertex positions for all of them.

**Comment by Alexustas** [comment]
> a rotateVariable that would rotate all of the line strings around 'position',

yes, PLEASE!

**Comment by MOARdV** [comment]
Okay.  Give me a few minutes. :)

**Comment by MOARdV** [comment]
Okay.  Beta updated with 'rotation' variable.  I will update the wiki now.

**Comment by MOARdV** [comment]
> Any chance of getting the ability to draw a custom dotted line?

I think we can attach a texture to the Unity LineRenderer material, but I have not tested it.  I think that will be the best way to add dotted lines if the LineRenderer object does not have special line modes.

**Comment by Alexustas** [comment]
https://youtu.be/AvIvuO3xeJg

**Comment by Alexustas** [comment]
looks like a one part of the line is drawn like this

![bandicam 2017-11-08 21-35-19-098](https://user-images.githubusercontent.com/6713671/32560974-d779a5d4-c4cc-11e7-8d84-c76c9ad8c820.jpg)

**Comment by MOARdV** [comment]
> looks like a one part of the line is drawn like this

Yes, I see that also.  It looks like it is a Unity LineRenderer bug that is fixed in Unity 5.5, but KSP is still Unity 5.4.

**Comment by Alexustas** [comment]
ok, I get it

**Comment by MOARdV** [comment]
Updated beta and wiki again :)

```
texture = MOARdV/MFD/dotted
```

Texture pattern is along the X coordinate of the texture (for instance, make a 32 x 4 texture).  MAS will try to keep the texture 1:1 (so, 32 pixel wide texture will repeat every 32 pixels of distance along the line), but another LineRenderer bug makes that not perfect if different sections of the line have different length.

**Comment by Alexustas** [comment]
> but another LineRenderer bug makes that not perfect if different sections of the line have different length.

yes, I can confirm this

https://youtu.be/twgxG_8jIgI

**Comment by MOARdV** [comment]
Both bugs are supposed to be fixed in newer versions of Unity, and I think KSP 1.4 will update Unity, so these will be better someday.

**Comment by Alexustas** [comment]
Is it possible to add the ability to crop the drawing region of a line?

Plan "A" (ideal scenario):

![bandicam 2017-11-14 06-29-52-464](https://user-images.githubusercontent.com/6713671/32758266-544126b4-c905-11e7-8a0a-41ebbc9b442b.jpg)

the rectangle is described by two vertices: the upper left corner and the right-bottom corner. vertices are set by offset from the "position". And the most important thing - cropping must be done AFTER rotation.

this can be done?

If not, Plan "B". 
we specify the size of the rectangle, it is centered on the "position"

but I really hope for the plan "A"))

**Comment by MOARdV** [comment]
Added `ELLIPSE`, which can also draw circles, of course.  Right now radius is not a variable, but I think it is easy to change to a variable.

Crop region will be a challenge.  I do not want to have to edit the line string every time a variable updates, because that will cause lag.

The best way probably is four black (or other color) rectangles that are drawn above the LineRenderer to hide the lines.  It will also hide anything else that is drawn below it.  Maybe a node named `VIEWPORT`?

**Comment by Alexustas** [comment]
ok. at least get rid of the extra large textures with the alpha channel. so let there be a Plan "C" )

**Comment by MOARdV** [comment]
Okay, Plan "C" - something like this:
```
VIEWPORT
{
  name = some name
  upperLeftCorner = x1, y1
  lowerRightCorner = x2, y2
  color = 255, 0, 0, 255
}
```
Will draw a viewport between upper-left and lower-right corner, and block everything outside of that viewport.

If work is not too busy, I will be able to make this today.

**Comment by MOARdV** [comment]
On [DropBox](https://www.dropbox.com/s/5lrxoiyjk96su34/AvionicsSystems-0.11.0b1.zip?dl=0).

As described above, and also supports `variable` and `range`.  All color channels and the corners can be variables.  Default color is black (0, 0, 0, 255).

**Comment by Alexustas** [comment]
> Added ELLIPSE, which can also draw circles, of course. Right now radius is not a variable, but I think it is easy to change to a variable.

Will it be difficult to draw arcs?

for example: 
```
drawArc = yes, 
startAngle = 270 (or variable)
endAngle = 180 (or variable)
longPath = yes
```

![bandicam 2017-11-15 21-43-44-320](https://user-images.githubusercontent.com/6713671/32848411-3e9b6776-ca4e-11e7-89af-ea21406dc255.jpg)

**Comment by MOARdV** [comment]
> Will it be difficult to draw arcs?

I finished testing `ELLIPSE`, and I asked myself, "Will alexustas want arcs?  Maybe if I do not add them, he will not notice?" :)

Yes, I can add arcs, and it is not very difficult.  It will work a little bit differently than your request, I think.  It will be easier to wind the same direction, so example 1 would be
```
startAngle = 270
endAngle = 180
```
and example 2 will be 
```
startAngle = 180
endAngle = 270
```
but we will see when I start working on it.

**Comment by Alexustas** [comment]
this is not a "MUSTHAVE" feature. at the moment I see an application for this feature only in that I can do this design of information representation

![bandicam 2017-11-16 03-46-36-202](https://user-images.githubusercontent.com/6713671/32864536-5ee920b2-ca81-11e7-9885-b428b050bf95.jpg)

 instead of this one's design

![bandicam 2017-11-16 03-46-44-799](https://user-images.githubusercontent.com/6713671/32864544-6935ab3a-ca81-11e7-859f-4fea3748b298.jpg)

and nothing more, but who knows ...   :-)

**Comment by Alexustas** [comment]
I bombards you with requests, because I need to know the maximum capabilities of MAS, so I can build a kind of strategy of development of the future IVAs. so not to repeat the history of RPM, when, as new features were added, I had to partially or completely rework all IVAs and props.

The key element of the next generation of my IVAs will be MFD. I want to minimize the number of dedicated single-function indicators and make cockpits as "glass" as possible. so I want the MFDs to have the maximum graphics capabilities and be versatile and flexible.
